src/RamInterface.cpp

```diff
@@ -0,0 +1,120 @@
+/* * Souffle - A Datalog Compiler
+* Copyright (c) 2016, The Souffle Developers. All rights reserved
+* Licensed under the Universal Permissive License v 1.0 as shown at:
+* - https://opensource.org/licenses/UPL
+* - <souffle root>/licenses/SOUFFLE-UPL.txt
+*/
+
+#include "RamInterface.h"
+
+namespace souffle {
+
+void RamRelationInterface::iterator_base::operator++() {
+    ++it;
+}
+
+tuple& RamRelationInterface::iterator_base::operator*() {
+    tup.rewind();
+
+    // construct the tuple to return
+    for (size_t i = 0; i < ramRelationInterface->getArity(); i++) {
+        if (*(ramRelationInterface->getAttrType(i)) == 's') {
+            std::string s = ramRelationInterface->getSymbolTable().resolve((*it)[i]);
+            tup << s;
+        } else {
+            tup << (*it)[i];
+        }
+    }
+
+    tup.rewind();
+
+    return tup;
+}
+
+RamRelationInterface::iterator_base* RamRelationInterface::iterator_base::clone() const {
+    return new RamRelationInterface::iterator_base(getId(), ramRelationInterface, it);
+}
+
+bool RamRelationInterface::iterator_base::equal(const Relation::iterator_base& o) const {
+    try {
+        auto iter = dynamic_cast<const RamRelationInterface::iterator_base&>(o);
+        return ramRelationInterface == iter.ramRelationInterface && it == iter.it;
+    } catch (const std::bad_cast& e) {
+        return false;
+    }
+}
+
+/**
+ * Helper function to convert a tuple to a RamDomain pointer
+ */
+RamDomain* convertTupleToNums(const tuple& t) {
+
+    std::vector<RamDomain> tuple;
+
+    for (size_t i = 0; i < t.size(); i++) {
+        tuple.push_back(t[i]);
+    }
+
+    return tuple.data();
```

Comment:
The internal array of the vector will not exist once you exit this function. The comments in tuple also advise against using the array notation to access it. I *think* you can do something like
```
RamDomain* tuple = new RamDomain[t.size()];
t.rewind();
for (...) {
  t >> tuple[i];
}
return tuple;
```
Otherwise, just use a new array instead of a vector in your existing code.